blog/2023/12/julia-1.10-highlights.md

```diff
@@ -0,0 +1,120 @@
++++
+mintoclevel = 2
+maxtoclevel = 3
+title = "Julia 1.10 Highlights"
+authors = "The Julia contributors"
+published = "22 Dec 2023"
+rss_pubdate = Date(2023, 12, 22)
+rss = """Highlights of the Julia 1.10 release."""
++++
+
+
+After X betas and X release candidates, Julia version 1.10 has finally(!) been released. We would like to thank all the contributors to this release and all the testers that helped with finding regressions and issues in the pre-releases. Without you, this release would not have been possible.
+
+The full list of changes can be found in the [NEWS file](https://github.com/JuliaLang/julia/blob/release-1.10/NEWS.md), but here we'll give a more in-depth overview of some of the release highlights.
+
+\toc
+
+# New parser written in Julia
+
+*Claire Foster*
+
+In Julia 1.10 we have replaced Julia's default parser (written in Scheme) with one written in Julia...
+
+# Package load time improvements
+
+*Jameson Nash et al.*
+
+In 1.9 the TTFX was heavily improved by allowing package authors to enable saving of native code during precompilation.
+
+In 1.10, as a follow up, signficant work was put into the performance of the loading of packages
+
+A lot of this work was driven by profiling and improving the load time of [OmniPackage.jl](https://github.com/JuliaComputing/OmniPackage.jl) which is an artifical "mega package" which only purpose is to depend on and load a lot of dependencies.
+
+- Improvement to the type system... (Jameson)
+- Reudction of invalidations causing recompilation
+- Moving packages away from Requires.jl to package extensions
+- `mul!` dispatch improvements (Daniel K)
+
+Omnipackage load times:
+
+Julia 1.9: 48.041773 seconds (102.17 M allocations: 6.522 GiB, 5.82% gc time, 1.21% compilation time: 86% of which was recompilation)
+Julia 1.10: 19.125309 seconds (30.38 M allocations: 2.011 GiB, 11.54% gc time, 10.38% compilation time: 61% of which was recompilation)
+
+# Stacktrace rendering improvements
+
+*Jeff Bezanson, Tim Holy*
+
+- `Type{...}` in stacktraces
+- Successive frames at identical location collapsed
+- https://github.com/JuliaLang/julia/pull/49102
+
+
+# Parallel GC
+
+*Diogo Netto*
+
+Up to 1.9, Julia's garbage collector was stop-the-world (meaning that all compute threads had to be stopped while a collection cycle was running) and serial (meaning a single thread performed a collection while the compute threads were halted). As one may imagine, garbage collection can quickly become a bottleneck in multithreaded code, particularly in code where multiple threads allocate.
+
+In 1.10, we gave the first steps towards improving scalability of the garbage collector in multithreaded code. More specifically, our work consisted of parallelizing the mark-phase of Julia's GC and enabling part of sweeping to be run concurrently with compute threads.
+
+# Tracy and Intel VTune ITTAPI profiling integration
```

Comment:
These features aren't (yet) enabled in our released images, but they are available to developers via build flags.

Do we still want to highlight this? Or should we wait until something like https://github.com/JuliaLang/julia/pull/50604 lands to make this available to everyone?

Reply:
I personally think it is still worth highlighting, perhaps with a view towards enabling it for all users in 1.11. It also ties in a bit with the load time improvements of packages in the other section in that having good profiling data made it a lot easier to figure out where work had to be done to improve the load times.